Src/IronPython/Runtime/Operations/StringOps.cs

```diff
@@ -1674,6 +1674,10 @@ public static class StringOps {
                 case "strict": e.DecoderFallback = final ? DecoderFallback.ExceptionFallback : new ExceptionFallBack(numBytes, e is UTF8Encoding); break;
                 case "replace": e.DecoderFallback = ReplacementFallback; break;
                 case "ignore": e.DecoderFallback = new PythonDecoderFallback(); break;
+                case "surrogateescape":
+                    e = (e.CodePage == 20127) ? // PythonAsciiEncoding insists on using its own fallbacks, downgrade to pure ASCII
```

Comment:
Is there a reason you need to do the check with `e.CodePage == 20127` instead of `e is PythonAsciiEncoding`? Adding `CodePage` to `PythonAsciiEncoding` is causing some stuff (e.g. `StringOps.GetEncodingName`) to fail.

Reply:
Checking for `e is PythonAsciiEncoding` will not work because `e` is actually `EncodingWrapper` that wraps `PythonAsciiEncoding`. The reason that things fail is because `EncodingWrapper` is not fully compliant with the `Encoding` protocol (i.e. it doesn't forward all public members to the wrapped encoding). Also `PythonAsciiEncoding` is not compliant.

Instead of comparing `CodePage` here I can compare `EncodingName`. I still will have to add a forwarder in `EncodingWrapper` but `PythonAsciiEncoding` already provides a proper name. It will be slightly less efficient as we will be comparing strings rather than integers.

Or I can make `EncodingWrapper` forward all calls to the wrapped encoding and make `PythonAsciiEncoding` call base with the ASCII code page number (this will make both compliant), but I cannot foresee all side consequences of this right now so it is maybe better to leave it as it was. In the future I would like to take a closer look at both classes; they seem like workarounds and maybe can be abolished or simplified.

Reply:
The only time `e` is an `EncodingWrapper` is when the encoding has a preamble. This should never happen with `PythonAsciiEncoding`.

Reply:
I see. OK then.